docs/HowToFAQ/FAQLoopModifies.md

```diff
@@ -0,0 +1,58 @@
+
+If that’s all the verifier did, life would be a pain, because you’d have to write a loop invariant about all variables in your program, even those that have nothing to do with the loop.
+Verifier actually does better. Instead of havocing all variables, it suffices to havoc the assignment targets within the body of the loop, including anything that might be modified by methods called in the loop body. That is, the verifier uses a simple syntactic scan of the loop body to see which variables may possibly be assigned, and then it havocs only those variables. This saves users from having to write loop invariants about all other variables. Only invariants about variables
+that are modified are needed.
+
+What about the heap? For this purpose, the heap is considered to be one variable. So, the heap is either an assignment target of the loop or it isn’t. This means that if your loop body assigns to anything in the heap, the heap becomes an assignment target. Also, if the loop body allocates another object, that too is a change of the heap, so the heap becomes an assignment target. Furthermore, Dafny’s rule about a method is that a method is allowed to change the state of any object in the method’s modifies clause, and it is also allowed to allocate new objects and to modify their state. Therefore, if your loop body calls a method, then the heap my well become an assignment target.
+ So, then, does this mean your loop invariant needs to speak about everything in the heap whenever the heap is an assignment target of the loop?
+
+Again, no. Loops in Dafny have `modifies` clauses, just like methods do (and just like for methods, newly allocated objects are implicitly included in those `modifies` clauses). It would be a pain to even have to write `modifies` clauses for every loop, so if a loop does not have an explicit `modifies` clause, Dafny implicitly attaches the `modifies` clause of the enclosing method (or of the enclosing loop, in the case of nested loops). This default behavior turns out to be right in almost all cases in practice, which is why you normally don’t have to think about this issue.
+
+But when it is not, you need to be explicit about the `modifies` clause and the loop invariants. In particular
+- write a `modifies` clause for the loop that includes all objects whose fields might be assigned in the loop body (or listed in the modifies clauses of anything the loop body calls)
+- then write a loop invariant that includes assertions about any variable that listed in the modifies clause. Very typically, the invariant will give a value for a havoced variable showing its relationship to the loop index.
+For example, a loop that sets the elements of an array to some initial value might loop like this:
+```dafny
+method init(a: array<int>) 
+  modifies a
+  ensures forall j: int :: 0 <= j < a.Length ==> a[j] == j;
```

Comment:
```suggestion
  ensures forall j | 0 <= j < a.Length :: a[j] == j;
```

(There are other examples of this pattern, which I think would be worth fixing too)